src/types.d.ts

```diff
@@ -0,0 +1,98 @@
+// Drift API types
+declare class Drift {
+  invoked: boolean
+  SNIPPET_VERSION: string
+  api: DriftAPI
+  allFramesReady: boolean
+  apiReady: boolean
+  controllerReady: boolean
+  hasInitialized: boolean
+  chatReady: boolean
+  config(config: DriftWidgetConfig): void
+  hide(): void
+  load(widgetId: string): void
+  show(): void
+  reset(): void
+  page(name: string): void
+  identify(userId: string, attributes: DriftUser): void
+  track(event: string, attributes: Record<string, unknown>[]): void
+  on(event: string, callback: DriftCallback | DriftEventCallback): void
+}
+
+// https://devdocs.drift.com/docs/widget-configuration
+type DriftWidgetConfig = {
+  locale?: string
+  messages?: {
+    welcomeMessage?: string
+    awayMessage?: string
+    emailCaptureMessage?: string
+    thankYouMessage?: string
+  }
+  enableWelcomeMessage?: boolean
+  enableCampaigns?: boolean
+  enableChatTargeting?: boolean
+  backgroundColor?: string
+  foregroundColor?: string
+  autoAssignee?: {
+    name: string
+    title: string
+    email: string
+    avatarUrl?: string
+  }
+  inboxId?: number
+  cookieDomain?: string
+  disableNewConversations?: boolean
+}
+
+type DriftCallback = (api: DriftAPI, payload?: Record<string, unknown>) => void
+
+type DriftEventCallback = (data: Record<string, unknown>[]) => void
```

Comment:
this is really not good and will always result in an any type for the first parameter.. `drift.on` can be either an event `ready` which will return the api and payload or it can be of any other event which will return the event data. How can we get this working? i guess overloads would be one way @mariux ?

Reply:
example:

```
drift.on('ready',function(api, payload){
  // hide the widget when it first loads
api.widget.hide()
 // show the widget when you receive a message
  drift.on('message',function(e){
     api.widget.show()
  })
  // hide the widget when you close the chat
    drift.on('chatClose',function(e){
     api.widget.hide()
  })
})
```

https://devdocs.drift.com/docs/widget-start

Reply:
will give it another try tomorrow once i slept with a clear mind

Reply:
yes, if you want this to be aligned, you need to define the different events possible and align them with the callbacks.. 

so its `on(event: 'ready', callback: DriftCallback): void` and `on(event: 'message' | 'chatClose', callback: DriftEventCallback): void` + all other events

Reply:
kk will do

Reply:
wow, the downloaded js file is 18K! lines long and even loads stuff like sentry Oo

https://js.driftt.com/include/1611058200000/3hczfxrfzggn.js

unfortunately, i am unable to generate types from this file :(

Reply:
implement all relevant methods and events strongly typed